internal/pkg/client/cache/dir.go

```diff
@@ -22,75 +46,268 @@ const (
 )
 
-var root string
+// SingularityCache is an opaque structure representing a cache
+type SingularityCache struct {
+	// BaseDir is the cache base directory. This directory is never entirely
+	// deleted since it can be specified by the user via the DirEnv environment
+	// variable.
+	BaseDir string
 
-// Root is the root location where all of singularity caching happens. Library, Shub,
-// and oci image formats supported by containers/image repository will be cached inside
-//
-// Defaults to ${HOME}/.singularity/cache
-func Root() string {
-	updateCacheRoot()
+	// rootDir is the cache root directory, within basedir. This is the directory
+	// Singularity actually manages, i.e., that can safely be deleted as
+	// opposed to the base directory that is potentially managed (passed in)
+	// by the user.
+	rootDir string
+
+	// State of the cache. We enable manual change of the state mainly for
+	// testing.
+	State string
+
+	// Default specifies if the handle points at the default image cache or
+	// not. This enables quick lookup.
+	Default bool
+
+	// PreviousDirEnv stores the value of the DirEnv environment variable
+	// before it is implicitly modified. This is used to restore the
+	// environment configuration in some specific contexts, and is therefore
+	// not meant to be exposed to users.
+	previousDirEnv string
+
+	// Library provides the location of the Library cache
+	Library string
+
+	// OciTemp provides the location of the OciTemp cache
+	OciTemp string
+
+	// OciBlob provides the location of the OciBlob cache
+	OciBlob string
+
+	// Net provides the location of the Net cache
+	Net string
+
+	// Shub provides the location of the Shub cache
+	Shub string
+}
+
+// NewHandle creates a new Singularity cache handle that can then be used to interact with a cache. The location of the cache is driven by the DirEnv directory. If it points to a location where a cache is not already present, a new cache will implicitly be created; otherwise, the new cache handle will point at the existing cache.
+func NewHandle() (*SingularityCache, error) {
+	// Singularity makes the following assumptions:
+	// - the default location for caches is specified by RootDefault
+	// - a user can specify the environment variable specified by DirEnv to change the location
+	// - a user can change the location of a cache at any time
+	// - but in the context of a Singularity command, the cache location cannot change once the command starts executing
+	basedir, err := getCacheBasedir()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get root of the cache: %s", err)
+	}
+
+	return hdlInit(basedir)
+}
+
+// Init initializes a new cache within a given directory
+func hdlInit(baseDir string) (*SingularityCache, error) {
+	rootDir := getCacheRoot(baseDir)
+	if err := initCacheDir(rootDir); err != nil {
+		return nil, fmt.Errorf("failed initializing caching directory: %s", err)
+	}
+
+	newCache := new(SingularityCache)
+	savedDirEnv := os.Getenv(DirEnv)
```

Comment:
It creates a new cache handle and save the environment variable associated to it (we could save `BaseDir`) so it can restored later on with the idea that meanwhile, in case it is internally changed, for example in the context of testing. I will add comments about this.

Reply:
If I understand correctly what you are saying, you are trying to protect against this:

```
$ cat test.go
package main

import (
        "fmt"
        "os"
)

func main() {
        oldfoo := os.Getenv("FOO")
        os.Setenv("FOO", "new value "+oldfoo)
        newfoo := os.Getenv("FOO")
        fmt.Printf("oldfoo=%q\nnewfoo=%q\n", oldfoo, newfoo)
}
```

A process cannot modify the environment of its parent. If you run the above program, you'll see:

```
$ export FOO="old value"
$ echo $FOO
old value
$ go run test.go
oldfoo="old value"
newfoo="new value old value"
$ echo $FOO
old value
```

If this is _not_ what you are saying, then obviously I still don't understand what this code is supposed to be doing. If this code is really necessary, it should have a very clear comment explaining what is its purpose.

To try to explain what it is that I don't understand: ideally, upon creation of the cache structure, the code should look at the environment exactly _once_. If the variable is set, it should use that value. If the variable is not set, it should use the default. From that point on, the rest of the code should either use the value for the cache base directory set in the cache structure or it should use whatever is passed as arguments to functions. In particular, it should never look at the environment variable again. With that in mind, what I don't understand is why the code should be modifying the environment, much less saving old values.

Reply:
No I do not try to prevent this, I know how environment variables are handled in the context of processes. I try to prevent the situation where a complex test is composed of two parts, one focusing on a higher-level capability which ultimately uses a cache under the cover. Furthermore, I want to make sure that I can "trick" the configuration of the cache to reach corner cases. All this in the context of a single unit test, i.e., a single process. So the high-level part of the test might set an environment variable (it is what privilege functions such as `DropPrivilege()` actually do) and once we reach the portion of the code dealing with the cache, the entire cache mechanism will deal with all the detail to create a temporary cache from scratch in a safe manner.